File: backend/endpoints/blogs.py
from fastapi import APIRouter, Header, Response
from utils.database import DatabaseAPI
from utils.objects import Blog
from typing import Annotated
import os
from time import time

router = APIRouter()


@router.post("/new")
async def new_blog(blog: Blog, token: Annotated[str | None, Header()], response: Response):
    if os.getenv("KEY") != token:
        response.status_code = 400
        return {"message": "Invalid Token"}
    blog.date = round(time())
    err = await DatabaseAPI.insert("blogs", vars(blog), blog.url)
    if err:
        response.status_code = err["code"]
        return {"message": err["message"]}
    return {"message": "Success"}


@router.get("/blogs")
async def get_blogs():
    return Blog.build(await DatabaseAPI.get("blogs", {}, "blogs"))


@router.get("/blogs/{id}")
async def get_blog(id):
    blog = Blog.build(await DatabaseAPI.get("blogs", {"url": id}, id))
    if len(blog) > 0:
        return blog[0]
    return {}

File: backend/endpoints/blogs.go
package endpoints

import (
	"encoding/json"
	"net/http"
	"os"

	"github.com/gorilla/mux"
	"osa-is.online/backend/utils"
)

func clean(bytes []byte) []byte {
	if bytes == nil {
		return []byte("[]")
	}
	return bytes
}

func Blogs(w http.ResponseWriter, r *http.Request) {
	res := utils.GetBlogs()
	w.Write(clean(res))
}

func NewBlog(w http.ResponseWriter, r *http.Request) {
	var blog utils.Blog
	headers := r.Header
	if headers.Get("Authorization") != os.Getenv("KEY") {
		http.Error(w, "Invalid Key", http.StatusForbidden)
		return
	}
	err := json.NewDecoder(r.Body).Decode(&blog)
	if err != nil {
		http.Error(w, "Malformed body", http.StatusBadRequest)
		return
	}
	utils.NewBlog(blog)
	w.Write([]byte("Added to Database"))
}

func Blog(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	id := vars["id"]
	res := utils.GetBlog(id)
	w.Write(clean(res))
}
